### ISSUE TITLE
Offline mode

### ISSUE THREAD

**Issue by orschiro** [mps-youtube/mps-youtube#733]
I know you can download music to listen offline with an external app but I actually got so much used to the great mpsyt interface that I would prefer to directly listen to the downloaded music through mpsyt while offline.

Are there any other users with a similar need or desire?

Thank you!

**Comment by ritiek** [comment]
A rough implementation in #737.

**Comment by orschiro** [comment]
@ritiek love your idea! Thanks so much. :-)

Should this local folder default to `~/Downloads/mps` which is where mps downloads songs to?

**Comment by Razesdark** [comment]
How about we tie this functionality to the playlist system instead, and take a spotify-like approach to offline media, where the user can enable *offline-mode* for one or more playlists and the software will handle caching these files locally for local playback. I think this approach is more in line with the intended functionality of mpsyt than blanket mp3 playback functionality. 

Although, a mpsyt like music file management solution sounds pretty sweet as well!

**Comment by orschiro** [comment]
@Razesdark nice idea. Maybe mpsyt can automatically create an offline playlist if the user doesn't specify a dedicated one.

**Comment by ritiek** [comment]
@Razesdark That sounds good idea too!

**Comment by ritiek** [comment]
@Razesdark That sounds a good idea too!

I think we'll have write the metadata to a text file or embed it in the music file itself so we can work with track duration (necessary to show playback progress when playing some song) and other metadata without internet.

**Comment by Razesdark** [comment]
Yeah, I was thinking something along the same lines! I got exams coming up on the 19th and 20th so I can't work on this until I'm done with those. So if anyone wants to start on this before then, feel free.

How about we extend the playlist to hold an offline flag. Then extend the video element to hold a *local path* string. If a video has a local path, we send that to mpv instead of querying the youtube api for a web url. (Maybe implement two function calls in in the video, one that returns a boolean describing if  the file is on disc, then a new function that returns the local path if the file is downloaded, and ytid if not.

This should make it pretty straight forward to adapt the play functions to use local files when available, and keep the commands free of Video related logic. Those files are bloated enough as it is IMO.

https://github.com/mps-youtube/mps-youtube/blob/develop/mps_youtube/playlist.py

How do we deal with for example caching a video, then setting show_video false. Should we only cache audio or maybe set the offline flag on a setting to setting basis (for example different settings for `show_video=true|false`, maybe transcoding options etc.)

Any other cases we need to consider?

**Comment by kraetzin** [comment]
@ritiek the metadata is already available in the playlist .m3u files. For example:

```
#EXTINF:215,Indila - Dernière Danse (Clip Officiel)
https://www.youtube.com/watch?v=K5KAc5CoCuk
```

You can see we have the duration and name of the track.

We can create custom metadata telling mpsyt that there is a local file, and where it is. For example, something like:
```
#EXTINF:215,Indila - Dernière Danse (Clip Officiel)
#MPSYTINF:local,~/mps/local_music_file.mp3
https://www.youtube.com/watch?v=K5KAc5CoCuk
```

@Razesdark When reading the m3u, mpsyt could then do a check for each item in a playlist and see if there is a local file available, in which case it should attempt to use that first. This would prevent the need of a special "offline playlist" type.

**Comment by Razesdark** [comment]
Agreed.

So we need a metadata tag that we can place directly after the `#EXTM3U` tag to denominate if mpsyt should treat that playlist as an offline playlist, and use the `#MPSYTINF` tag @kraetzin suggested to store the path for each respective file in the playlist.

How about `#MPSYTINF:offline,audio|video` near the top of the file to tell mpsyt if it should cache audio or video versions of the playlist. Not having a tag would mean that it is not an offline playlist. The user would type a command similar to `cache pl n audio` or `cache pl n video` when they designate a playlist as offline

**Comment by kraetzin** [comment]
That sounds good. The `cache` command could also default `audio` or `video` to whatever `show_video` says.

Looking into the issue of having a video file cached and then setting `show_video=false`, `mpv` has a `--no-video` option which could be set. In the case of `show_video=true` being set and attempting to play an audio file, just play the audio. I don't know about `mplayer` though.

In addition, we could be more explicit in the track info tags and specify `#MPSYTINF:local_audio,music_audio.mp3` and `#MPSYTINF:local_video,music_video.mp4` to accomodate both `cache pl n audio` and `cache pl n video` being called on one playlist, since we'd end up with 2 different files.

**Comment by iamtalhaasghar** [comment]
Related #993
